### PR TITLE
Add more type hints to Model and ModelInfo

### PR DESCRIPTION
**Pull request by agjendem**
### Related Issues/PRs

Related to #13138, adding more type hints to avoid having mypy complain that methods are untyped. 

Originally also typed `get_input_schema()` and `get_output_schema()` with `Optional[Schema]` from mlflow.types.schema.Schema, but as per https://github.com/mlflow/mlflow/blob/6913f67be07fc09e1bd1aa72f2f698cec56dd6e0/mlflow/models/model.py#L92 and Schema that would require Numpy to be a dependency.

### How is this PR tested?

- [X] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [X] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [X] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [X] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

**Comment by serena-ruan** [comment]
@agjendem Could you add `Self` to https://github.com/mlflow/mlflow/blob/6fa504db336d1d3135a1953cc236592652f02906/docs/source/conf.py#L411 to avoid breaking build_doc job?

File: mlflow/models/model.py [hunk 1]
```diff
@@ -343,24 +344,24 @@ def __eq__(self, other):
 
     def get_input_schema(self):
         """
-        Retrieves the input schema of the Model iff the model was saved with a schema definition.
+        Retrieves the input schema of the Model if the model was saved with a schema definition.
```
Comment (serena-ruan) [inline mlflow/models/model.py:347]: Let's revert these `iff -> if` changes

**Review by serena-ruan — APPROVED** [review]
Overall LGTM!

File: mlflow/models/model.py [hunk 2]
```diff
@@ -12,6 +12,7 @@
 
 import yaml
 from packaging.requirements import InvalidRequirement, Requirement
+from typing_extensions import Self
```
Comment (harupy) [inline mlflow/models/model.py:15]: does mlflow have typing_extensions as a dependency?

Comment (agjendem) [reply mlflow/models/model.py:15]: Based on the Python 3.8 pip environment I got when running `dev/dev-env-setup.sh -d .venvs/mlflow-dev -q` from the contributions guide + querying with `pipdeptree`: `pipdeptree --reverse --packages typing_extensions`
 it does not seem to be a _explicit_ dependency in core-requirements.txt, but it is a direct dependency of both `alembic` and `graphene`, that is listed.

Comment (harupy) [reply mlflow/models/model.py:15]: We should not use a package that mlflow doesn't depend on directly. Can we just a forward reference?

Comment (agjendem) [reply mlflow/models/model.py:15]: It seems we can, wasn't aware of that syntax. Thanks for the tip!

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**